### PR TITLE
Issue 5804: Optimized manifest file size

### LINKED ISSUE #5804: Optimize the manifest format
For a small sized project, the `.dmanifest` is disproportinally large (67% of the actual content!)
The main purpose of the manifest is to keep a list of signatures of files, and also keep a list of file dependants

```
-rw-r--r--  1 mawe  staff  14057322 May  7 18:53 build/default/game.arcd
-rw-r--r--  1 mawe  staff   3072208 May  7 18:53 build/default/game.arci
-rw-r--r--  1 mawe  staff   9487653 May  7 18:53 build/default/game.dmanifest
-rw-r--r--  1 mawe  staff      4700 May  7 18:53 build/default/game.projectc
-rw-r--r--  1 mawe  staff       162 May  7 18:53 build/default/game.public.der
```

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/liveupdate/src/liveupdate_private.cpp [hunk 1]
```diff
@@ -36,16 +36,16 @@ namespace dmLiveUpdate
         while (first <= last)
         {
             int mid = first + (last - first) / 2;
-            uint64_t currentHash = entries[mid].m_UrlHash;
-            if (currentHash == urlHash)
+            dmhash_t currentHash = entries[mid].m_UrlHash;
```
Comment (JCash) [inline engine/liveupdate/src/liveupdate_private.cpp:39]: Cleanup (correct type, and snake_case)

File: engine/liveupdate/src/liveupdate_private.cpp [hunk 2]
```diff
@@ -54,34 +54,52 @@ namespace dmLiveUpdate
         return NULL;
     }
 
-    uint32_t MissingResources(dmResource::Manifest* manifest, const dmhash_t urlHash, uint8_t* entries[], uint32_t entries_size)
+    uint32_t MissingResources(dmResource::Manifest* manifest, const dmhash_t url_hash, uint8_t* entries[], uint32_t entries_size)
     {
         uint32_t resources = 0;
 
-        if (manifest == 0x0)
+        if (manifest == 0x0 || entries == 0x0 || entries_size == 0)
```
Comment (JCash) [inline engine/liveupdate/src/liveupdate_private.cpp]: No need to check these in the for loop below

File: engine/liveupdate/src/liveupdate_private.cpp [hunk 3]
```diff
@@ -54,34 +54,52 @@ namespace dmLiveUpdate
...
+        HResourceEntry entry = FindResourceEntry(manifest, url_hash);
         if (entry != NULL)
         {
             for (uint32_t i = 0; i < entry->m_Dependants.m_Count; ++i)
             {
-                uint8_t* resource_hash = entry->m_Dependants.m_Data[i].m_Data.m_Data;
-                dmResourceArchive::Result result = dmResourceArchive::FindEntry(manifest->m_ArchiveIndex, resource_hash, hash_len, 0, 0);
-                if (result != dmResourceArchive::RESULT_OK)
+                dmhash_t dependant_url_hash = entry->m_Dependants.m_Data[i];
+                HResourceEntry dependant = FindResourceEntry(manifest, dependant_url_hash);
```
Comment (JCash) [inline engine/liveupdate/src/liveupdate_private.cpp]: use the shorter (8 byte) url_hash, vs the cryptographic hash (40 bytes for sha256) to find the resource entry.

File: engine/liveupdate/src/liveupdate_private.cpp [hunk 4]
```diff
@@ -66,26 +64,41 @@ namespace dmLiveUpdate
...
+        {
+            return count;
+        }
+
+        uint32_t num_resources = 0;
+        for (uint32_t i = 0; i < count; ++i)
         {
-            for (uint32_t i = 0; i < entry->m_Dependants.m_Count; ++i)
+            dmhash_t dependant_url_hash = entry->m_Dependants.m_Data[i];
+            HResourceEntry dependant = FindResourceEntry(manifest, dependant_url_hash);
```
Comment (JCash) [inline engine/liveupdate/src/liveupdate_private.cpp:84]: Use the shorter hash, to find the same resource entry.

File: engine/resource/proto/liveupdate_ddf.proto [hunk 5]
```diff
@@ -99,7 +99,7 @@ message ResourceEntry {
     required string             url = 1;
     required uint64             url_hash = 2;
     required HashDigest         hash = 3;
-    repeated HashDigest         dependants = 4;
+    repeated uint64             dependants = 4;
```
Comment (JCash) [inline engine/resource/proto/liveupdate_ddf.proto:102]: The test project went from 9mb to 6mb (with a 14mb archive).

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java [hunk 6]
```diff
@@ -504,32 +520,69 @@ public ManifestHeader buildManifestHeader() throws IOException {
...
-            for (String dependant : dependants) {
-                for (ResourceEntry dependantEntry : this.resourceEntries) {
-                    if (dependantEntry.getUrl().equals(dependant)) {
-                        if (dependantEntry.hasHash()) {
-                            resourceEntryBuilder.addDependants(dependantEntry.getHash());
-                        } else {
-                            throw new IOException("Unable to create ManifestData, an incomplete resource was found!");
+            // Since we'll only ever ask collection proxies, we only store those lists
+            if (url.endsWith("collectionproxyc"))
+            {
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java:564]: This saves an extra ~2mb

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java [hunk 7]
```diff
@@ -504,32 +520,69 @@ public ManifestHeader buildManifestHeader() throws IOException {
...
+                // We'll only store the dependencies for the excluded collection proxies
+                if (excludedResources.contains(url)) {
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java:566]: This saves an extra ~400kb

**Review by britzl — DISMISSED** [review]
Looks good! :+1

**Pull request merged**